### PR TITLE
Fix handling of invalid utf-8 sequences in PyString::to_string_lossy

### PR DESCRIPTION
**Pull request by Alexander-N**
See discussion in #479 and #634.

### CODE REVIEW

File: src/types/string.rs [hunk 1]
```diff
@@ -87,10 +88,29 @@ impl PyString {
...
+                        self.0.as_ptr(),
+                        CStr::from_bytes_with_nul(b"utf-8\0").unwrap().as_ptr(),
+                        CStr::from_bytes_with_nul(b"surrogatepass\0")
+                            .unwrap()
+                            .as_ptr(),
+                    );
+                    // Since we have a valid PyString and replace any surrogates, assume success.
+                    debug_assert!(!py_bytes.is_null());
+                    // ensure DECREF will be called
+                    self.py().from_owned_ptr::<PyBytes>(py_bytes);
```
Comment (kngwyu) [inline src/types/string.rs]: How about `crate::gil::register_pointer`?(My snippet was incorrect, sorry)

**Comment by kngwyu** [comment]
Thanks!

File: src/types/string.rs (hunk 1 above)
Comment (Alexander-N) [reply src/types/string.rs]: Ok, done. What's the difference?

Comment (kngwyu) [reply src/types/string.rs]: 1. Less overhead.
2. Our GIL pool has 3 types of buffer(owned ref, borrowed ref, pointer). 
In this case, we can use both owned and pointer buffer.
But pointer buffer is preferable for the case `GILGuard` drops earlier(I'll add an example but have no time now :sweat: ).

**Pull request merged**